presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -484,6 +485,30 @@ public PlanWithProperties visitSort(SortNode node, Context context)
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
+            else {
+                // current plan so far is single node, so local properties are effectively global properties
+                // if the local properties guarantee ordering on Sort keys
+                // TODO: This should be extracted as a separate optimizer once the planner is able to reason about the ordering of each operator.
+                if (child.getProperties().getLocalProperties().size() >= node.getOrderBy().size()) {
+                    boolean childAlreadySorted = true;
+                    for (int i = 0; i < node.getOrderBy().size(); i++) {
+                        Symbol desiredColumn = node.getOrderBy().get(i);
+                        SortOrder desiredOrdering = node.getOrderings().get(desiredColumn);
+
+                        LocalProperty<Symbol> childProperty = child.getProperties().getLocalProperties().get(i);
+                        if (!(childProperty instanceof SortingProperty) ||
```

Comment:
You should be able to use `LocalProperties.match(...)` instead of rolling your own.